### ISSUE TITLE
document.hasFocus is incorrect if browser window is loaded/refreshed while not visible or unfocused

### ISSUE THREAD

**Issue by jwheare** [electron/electron#7125]
- Electron version: 1.3.5
- Operating system: macOS 10.11.6

`document.hasFocus` tells you whether the window is blurred or focus. If a BrowserWindow instance is loaded while hidden or unfocused, it will be set to true and it's not possible to blur it from that state.

I'm currently working around this by briefly showing/focusing the window, then blurring it from a preload script.

``` javascript
var remote = require('electron').remote;

function setupFocus () {
  // If we load/refresh while the window is hidden or blurred, document.hasFocus is wrong
  // We need to first show the window it it's hidden (this is probably ok if a user manually
  // chose to refresh) or give it focus (e.g. if dev tools is open) then blur it again to properly
  // set the document's focus state
  if (!remote.getCurrentWindow().isVisible() || !remote.getCurrentWindow().isFocused()) {
    if (!remote.getCurrentWindow().isVisible()) {
      remote.getCurrentWindow().show();
    } else {
      remote.getCurrentWindow().focus();
    }
    remote.getCurrentWindow().blur();
  }
}

module.exports = setupFocus;
```

**Comment by liusy182** [comment]
I am not sure if i follow correctly but I am not able to reproduce on 1.3.5. I was able to get correct state by calling `document.hasFocus()` after the window had been refreshed when it was in in blurred state.

**Comment by zcbenz** [comment]
I can reproduce by opening the default app and printing `document.hasFocus()` in devtools, it seems to always be `false`.

**Comment by jwheare** [comment]
@liusy182 might depend on what you mean by blurred state.

For me, it's having a devtools window open and focused (not just the app being in the background)

Also, try to refresh via the devtools Cmd-r shortcut, after calling browserWindow.hide(); from main process.

@zcbenz that sounds like the opposite, did you mean `true` or that you *can't* reproduce?

**Comment by MarshallOfSound** [comment]
> I can reproduce by opening the default app and printing document.hasFocus() in devtools, it seems to always be false.

@zcbenz This is caused by the devtools being in focus, if you run this code

```js
setTimeout(() => console.log(document.hasFocus()), 1000)
```

And immediately click back to the main window it will print `true` as expected.  I can't reproduce the original issue 😢

**Comment by jwheare** [comment]
## Steps to reproduce on macOS

1. Run electron default_app with `--app=https://jsfiddle.net/2prtLvjw/1/` and open https://jsfiddle.net/2prtLvjw/1/ in Chrome to compare.
2. Cmd-Alt-I to toggle dev tools and ensure they have focus
3. Run `document.hasFocus()` in devtools console again
4. Press Cmd-R to refresh the page via the devtools shortcut (not the Reload menu item, that will only reload the focused window, i.e. devtools
5. Run `document.hasFocus()` in devtools console again

## Expected behaviour (this is Chrome's behaviour)

1. After initial page load, devtools console log shows `true` and fiddle panel HTML shows `focus: true`
2. `document.hasFocus()` returns `false`
3. After page refresh with devtools focused, console log shows `false` and fiddle HTML shows `focus: false`
4. `document.hasFocus()` returns `false`

## Actual behaviour (1 and 2 match Chrome)

1. After initial page load, devtools console log shows `true` and fiddle panel HTML shows `focus: true`
2. `document.hasFocus()` returns `false`
3. After page refresh with devtools focused, console log shows `true` and fiddle HTML shows `focus: true`
4. `document.hasFocus()` returns `true`

If you change the lines in `default_app.js` at the end of the `load` function as below, steps 1 and 2 also become invalid.

1. After initial page load, devtools console log shows `true` and fiddle panel HTML shows `focus: true`
2. `document.hasFocus()` returns `true`

The initial load should not be focused, and the check from devtools console should *never* return true.

```javascript
    mainWindow = new BrowserWindow(options)
    mainWindow.hide()
    mainWindow.loadURL(appUrl)
    mainWindow.focus()
    mainWindow.toggleDevTools()```

**Comment by zcbenz** [comment]
Can you try v1.4.0? I'm unable to reproduce with it.

**Comment by jwheare** [comment]
Yup still happens on a fresh build.

Screenshot of initial load, check, refresh, check, all with devtools focused.

![screen shot 2016-09-19 at 20 35 15](https://cloud.githubusercontent.com/assets/49224/18646251/9d98444e-7ea8-11e6-852f-216b6406b20d.png)

**Comment by jwheare** [comment]
Happy to debug this with lldb if you can point me to an appropriate breakpoint.

**Comment by zcbenz** [comment]
I'm merging this to #6860 since they are essentially the same issue.

**Comment by jwheare** [comment]
Just got around to testing this with electron 1.8.1 and this bug is still happening despite #6860 being marked as fixed.

**Comment by codebytere** [comment]
is this still happening on the latest release in that line?

**Comment by jwheare** [comment]
Yup, this still happens with Electron 1.8.4.

**Comment by codebytere** [comment]
@jwheare gotcha, i'll check this out this week!

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v2.0.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons.

Thanks in advance! Your help is appreciated.

**Comment by jwheare** [comment]
Just tested in electron 4.0.1 with the latest electron-quick-start app. Updated simpler reproduction instructions:

## Steps to reproduce on macOS
1. `git clone https://github.com/electron/electron-quick-start`
2. `npm install`
3. `npm start`
4. Press `Cmd-Alt-I` to open dev tools
5. Run `document.hasFocus()` in devtools console
6. Press `Cmd-R` to refresh the page via the devtools shortcut (not the Reload menu item, that will only reload the devtools window)
5. Run `document.hasFocus()` in devtools console again

## Expected behaviour (this is Chrome's behaviour, try on any website)
✅ After loading the page and then opening devtools, `document.hasFocus()` returns `false`
✅ After page refresh **with devtools still focused** `document.hasFocus()` returns `false`

## Actual behaviour
✅ After loading the page and then opening devtools, `document.hasFocus()` returns `false`
❌ After page refresh **with devtools still focused** `document.hasFocus()` returns ---> `true` <---

**Comment by codebytere** [comment]
This should be fixed with latest v7, as verified with the following steps on macOS:

```
- git clone https://github.com/electron/electron-quick-start
- npm install
- npm start
- cmd-alt-I to open dev tools
- run document.hasFocus() in devtools console
- cmd-R to refresh the page via the devtools shortcut
- Run document.hasFocus() in devtools console again
```

**Comment by jwheare** [comment]
I'm afraid this is still not fixed in 8.0.0. To test this fully, make sure the devtools is open in a separate window, not just a side pane.

**Comment by jwheare** [comment]
Just tested this in 11.1.1 and it appears to now be fixed.
